### PR TITLE
Only load compose file when necessary

### PR DESCRIPTION
**Pull request by edwardrf**

### CODE REVIEW

File: src/pkg/cli/project/load.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+	"github.com/compose-spec/compose-go/v2/loader"
+	compose "github.com/compose-spec/compose-go/v2/types"
+	"github.com/defang-io/defang/src/pkg/term"
+	"github.com/defang-io/defang/src/pkg/types"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+var ComposeFilePath string
```
Comment (lionello) [inline src/pkg/cli/project/load.go:17]: move to `internal/`?

File: src/cmd/cli/command/commands.go [hunk 2]
```diff
@@ -290,34 +286,9 @@ var RootCmd = &cobra.Command{
...
-					term.Warn(" - Error loading docker compose file: ", err)
-				}
-			} else {
-				projectName = project.Name
-			}
-		}
-
-		client = cli.NewClient(cluster, projectName, provider)
+		project.ComposeFilePath, _ = cmd.Flags().GetString("file")
+		project.TenantID = cli.GetTenantID(cluster)
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:290]: we now have `WhoAmI` so we could probably remove this hack `cli.GetTenantID`

File: src/cmd/cli/command/commands.go [hunk 3]
```diff
@@ -706,15 +676,19 @@ func printEndpoints(serviceInfos []*defangv1.ServiceInfo) {
...
+	Annotations: authNeededAnnotation,
 	Args:        cobra.NoArgs, // TODO: takes optional list of service names
 	Short:       "Like 'start' but immediately tracks the progress of the deployment",
 	RunE: func(cmd *cobra.Command, args []string) error {
 		var force, _ = cmd.Flags().GetBool("force")
 		var detach, _ = cmd.Flags().GetBool("detach")
 
 		since := time.Now()
-		deploy, err := cli.ComposeStart(cmd.Context(), client, project, force)
+		proj, err := client.LoadCompose()
```
Comment (lionello) [inline src/cmd/cli/command/commands.go:687]: wonder if we can move this to the `composeCmd`? Would make a lot of sense!

**Review by lionello — APPROVED** [review]
just two comments

File: src/cmd/cli/command/commands.go (hunk 3 above)
Comment (edwardrf) [reply src/cmd/cli/command/commands.go]: Yeah, had the same idea, now moved to the respective Compose command.

File: src/pkg/cli/connect.go [hunk 4]
```diff
@@ -60,16 +60,15 @@ func Connect(cluster string) (*client.GrpcClient, types.TenantID) {
...
 
-	return client.NewGrpcClient(host, accessToken), tenantId
+	return client.NewGrpcClient(host, accessToken, tenantId, loader), tenantId
 }
 
-func NewClient(cluster string, projectName string, provider client.Provider) client.Client {
-	term.Debug(" - Project", projectName)
-	defangClient, tenantId := Connect(cluster)
+func NewClient(cluster string, provider client.Provider, loader client.ProjectLoader) client.Client {
+	defangClient, tenantId := Connect(cluster, loader)
```
Comment (edwardrf) [inline src/pkg/cli/connect.go:68]: Should I add a `defangClient.Whoami` here to validate the tenant?

File: src/cmd/cli/command/commands.go (hunk 2 above)
Comment (edwardrf) [reply src/cmd/cli/command/commands.go]: After restructuring, the TenantID is no longer needed here, but I think the right place for us to get the tenantID from fabric would be on connect.go @ L68, see my comments below.

Comment (edwardrf) [reply src/cmd/cli/command/commands.go]: I just added the tenant ID reloading, let me know what do you think

File: src/pkg/cli/composeRestart.go [hunk 5]
```diff
@@ -4,11 +4,14 @@ import (
...
 	"github.com/defang-io/defang/src/pkg/cli/client"
 )
 
-func ComposeRestart(ctx context.Context, client client.Client, project *composeTypes.Project) (client.ETag, error) {
+func ComposeRestart(ctx context.Context, client client.Client) (client.ETag, error) {
+	project, err := client.LoadProject()
+	if err != nil {
+		return "", err
+	}
 	if project == nil {
```
Comment (lionello) [inline src/pkg/cli/composeRestart.go:15]: Can this happen? `nil, nil`?

File: src/pkg/cli/composeStop.go [hunk 6]
```diff
@@ -4,11 +4,14 @@ import (
...
 	"github.com/defang-io/defang/src/pkg/cli/client"
 )
 
-func ComposeStop(ctx context.Context, client client.Client, project *composeTypes.Project) (client.ETag, error) {
+func ComposeStop(ctx context.Context, client client.Client) (client.ETag, error) {
+	project, err := client.LoadProject()
+	if err != nil {
+		return "", err
+	}
 	if project == nil {
```
Comment (lionello) [inline src/pkg/cli/composeStop.go:15]: ditto

File: src/pkg/cli/connect.go [hunk 7]
```diff
@@ -60,16 +61,24 @@ func Connect(cluster string) (*client.GrpcClient, types.TenantID) {
 	}
 	term.Debug(" - Using tenant", tenantId, "for cluster", host)
 
-	return client.NewGrpcClient(host, accessToken), tenantId
+	defangClient := client.NewGrpcClient(host, accessToken, tenantId, loader)
+	resp, err := defangClient.WhoAmI(context.Background()) // TODO: Should we pass in the command context?
```
Comment (lionello) [inline src/pkg/cli/connect.go:65]: use `context.TODO()` for now

File: src/pkg/cli/connect.go [hunk 8]
```diff
@@ -60,16 +61,24 @@ func Connect(cluster string) (*client.GrpcClient, types.TenantID) {
...
+	if err != nil {
+		term.Debug(" - Unable to validate tenant ID with server:", err)
+	}
+	if resp != nil && tenantId != types.TenantID(resp.Tenant) {
+		term.Warnf(" - Overriding locally cached TenantID %v with server provided value %v", tenantId, resp.Tenant)
```
Comment (lionello) [inline src/pkg/cli/connect.go:70]: nit: we use `!` for warn, `-` for debug

**Review by lionello — APPROVED** [review]
fantastic

File: src/pkg/cli/composeRestart.go (hunk 5 above)
Comment (edwardrf) [reply src/pkg/cli/composeRestart.go]: Good call, removed

**Pull request merged**